### PR TITLE
Security Manager example

### PR DESCRIPTION
**Pull request by paul-szczepanek-arm**
To demonstrate the basic use of the Security Manager and help validate the new features.

There are no tests for it yet, the ccli-app is being developed in parallel as it needs to provide access to the new API calls.

### CODE REVIEW

**Comment by paul-szczepanek-arm** [comment]
@pan- @donatieng please review

File: BLE_SM/README.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# SM - example usage of the Security Manager to pair and encrypt
+
+Demonstration of possible usage of the Security Manager. Security Manager deals with pairing, authentication and encryption.
+
+The application demonstrates usage as a central and a peripheral. The central will connect to any connectable device present. Please have one ready and advertising. Application will attempt pairing. Please authorise your peer device to pair.
+
+Upon success it will disconnect and start advertising to demonstrate usage as a peripheral. Please scan and connect using your peer device. Upon connection grant pairing if prompted. Upon success the application will disconnect. Observe the terminal to keep track of the sequence.
```
Comment (pan-) [inline BLE_SM/README.md]: Phase two will start whether phase 1 succeed or not. Could you indicate that if phase 1 fail, the peripheral will request pairing or if phase 1 succeeded then the link is encrypted.

File: BLE_SM/source/main.cpp [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+            printf("Link ENCRYPTED\r\n");
+        } else if (result == link_encryption_t::ENCRYPTED_WITH_MITM) {
+            printf("Link ENCRYPTED_WITH_MITM\r\n");
+        } else if (result == link_encryption_t::NOT_ENCRYPTED) {
+            printf("Link NOT_ENCRYPTED\r\n");
+        }
+    }
+
+protected:
+    /** Override to start chosen activity when initialisation completes */
```
Comment (pan-) [inline BLE_SM/source/main.cpp]: Is there a reason to have a protected block here ? private seems fine.

File: BLE_SM/source/main.cpp [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+        /* If the security manager is required this needs to be called before any
+         * calls to the Security manager happen. */
+        error = _ble.securityManager().init();
+
+        if (error) {
+            printf("Error during init %d\r\n", error);
+            return;
+        }
+
+        /* Tell the security manager to use methids in this class to inform us
```
Comment (pan-) [inline BLE_SM/source/main.cpp]: typo: methids => methods

File: BLE_SM/source/main.cpp [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+        _ble.gap().setAdvertisingTimeout(0);
+
+        error = _ble.gap().startAdvertising();
+
+        if (error) {
+            printf("Error during Gap::startAdvertising.\r\n");
+            return;
+        }
+
+        _ble.securityManager().setPairingRequestAuthorisation(true);
```
Comment (pan-) [inline BLE_SM/source/main.cpp]: Could you add a comment explaining what this means ?

File: BLE_SM/source/main.cpp (hunk 2 above)
Comment (paul-szczepanek-arm) [reply BLE_SM/source/main.cpp]: nope, removed

**Comment by paul-szczepanek-arm** [comment]
Will address comments and change the phases around so that we can store the address of the peer in the example app.

File: BLE_SM/README.md (hunk 1 above)
Comment (paul-szczepanek-arm) [reply BLE_SM/README.md]: reversed the order

File: BLE_SM/mbed-os.lib [hunk 5]
```diff
@@ -0,0 +1 @@
+https://github.com/paul-szczepanek-arm/mbed-os/tree/palsm
```
Comment (donatieng) [inline BLE_SM/mbed-os.lib]: Can you make it point to the 5.8 branch?

File: BLE_SM/mbed_app.json [hunk 6]
```diff
@@ -0,0 +1,16 @@
+{
+    "target_overrides": {
+        "K64F": {
+            "target.features_add": ["BLE"],
+            "target.extra_labels_add": ["ST_BLUENRG"]
+        },
+        "NUCLEO_F401RE": {
```
Comment (donatieng) [inline BLE_SM/mbed_app.json]: Please document that it is using the Cordio port for ST (as opposed to ST's official stack)

Comment (paul-szczepanek-arm) [reply BLE_SM/mbed_app.json]: note made about it in the readme

**Comment by pan-** [comment]
> Would it be good to add a configuration step where the device act as a peripheral and wait for a connection. The device which has initiated the connection will then be used in phase 1.

I've tried the amended example and it may not work if the phone advertise another address. 
We can use manufacturer data in advertising payload as a more reliable option.

File: BLE_SM/mbed-os.lib [hunk 7]
```diff
@@ -1 +1 @@
-https://github.com/ARMmbed/mbed-os/tree/mbed-os-5.8
+https://github.com/ARMmbed/mbed-os/tree/mbed-os-5.8.2
```
Comment (pan-) [inline BLE_SM/mbed-os.lib]: Could you use a commit ID ? So it's aligned with other example.

**Pull request merged**